dd-sdk-android-ndk/src/test/cpp/test-signal-monitor.cpp

```diff
@@ -32,59 +40,98 @@ bool performedInstall(int times) {
 }
 
-TEST calling_install_more_times_in_a_row_will_only_install_once(void) {
-    install_signal_handlers();
-    install_signal_handlers();
-    install_signal_handlers();
+TEST calling_start_monitoring_more_times_in_a_row_will_only_install_once(void) {
+    start_monitoring();
+    start_monitoring();
+    start_monitoring();
     // give time for the install thread to finish
-    sleep(5);
+    pthread_join(datadog_watchdog_thread, nullptr);
             ASSERT(handlers_installed);
             ASSERT(performedInstall(1));
-    uninstall_signal_handlers();
     clear_tests();
             PASS();
 }
 
 
-TEST uninstalls_signal_handlers(void) {
+TEST calling_stop_monitoring_will_uninstall_the_handlers(void) {
     // given
-    install_signal_handlers();
-    sleep(5);
+    start_monitoring();
+    pthread_join(datadog_watchdog_thread, nullptr);
     // when
-    uninstall_signal_handlers();
+    stop_monitoring();
             ASSERT_FALSE(handlers_installed);
     clear_tests();
             PASS();
 }
 
-TEST calling_uninstall_more_times_in_a_row_will_only_uninstall_once(void) {
+TEST calling_stop_monitoring_more_times_in_a_row_will_only_uninstall_once(void) {
     // given
-    install_signal_handlers();
-    sleep(5);
+    start_monitoring();
+    pthread_join(datadog_watchdog_thread, nullptr);
     // when
-    uninstall_signal_handlers();
-    uninstall_signal_handlers();
-    uninstall_signal_handlers();
+    stop_monitoring();
+    stop_monitoring();
+    stop_monitoring();
             ASSERT_FALSE(handlers_installed);
             ASSERT(performedUninstall(1));
     clear_tests();
             PASS();
 }
 
+TEST calling_start_monitor_from_different_threads_will_only_install_once(void) {
+    auto start_monitoring_lambda = [] {
+        start_monitoring();
+    };
+    std::thread t1 = std::thread(start_monitoring_lambda);
+    std::thread t2 = std::thread(start_monitoring_lambda);
+    t1.join();
+    t2.join();
+    pthread_join(datadog_watchdog_thread, nullptr);
+            ASSERT(handlers_installed);
+            ASSERT(performedInstall(1));
+    clear_tests();
+            PASS();
+}
+
+TEST calling_start_monitor_from_different_threads_will_not_override_the_sigaction(void) {
+    auto start_monitoring_lambda = [] {
+        start_monitoring();
+    };
+    std::thread t1 = std::thread(start_monitoring_lambda);
+    std::thread t2 = std::thread(start_monitoring_lambda);
+    t1.join();
+    t2.join();
+    pthread_join(datadog_watchdog_thread, nullptr);
+    int sigactions_size = 0;
+    const int handled_signals = 6;
+    struct sigaction *end_pointer = datadog_sigaction + handled_signals;
+    struct sigaction *pointer = datadog_sigaction;
+    while (pointer != end_pointer) {
+        sigactions_size++;
+                ASSERT(pointer != nullptr);
+        pointer++;
+    }
+            ASSERT(handlers_installed);
+            ASSERT(performedInstall(1));
+            ASSERT_EQ(sigactions_size, handled_signals);
```

Comment:
Have you seen this assertion fail ? Based on the code, it seems it can never fail, why are you checking this ? Also how are you asserting that the `sigaction` is not overriden when start is called twice ?

Reply:
Yes...I wanted to make sure. Basically you cannot check if it is overridden but what you can check is if it is in an inconsistent state. Basically what I wanted to test is :

1.  Thread t1 is starting (the memory allocation starts on the t1 thread and watchdog thread is started)
2. While watchdog is working the t2 starts say `watchdog` is still working then t2 will try to allocate new memory. Without the fence this will compromise the memory allocated in t1 and will probably crash. I guess you are actually right in the sense that this test should be renamed but I would like to keep the assertions to make sure my locks are working.